src/analysis/normed_space/basic.lean

```diff
@@ -596,6 +588,141 @@ lemma nat.norm_cast_le [has_one α] : ∀ n : ℕ, ∥(n : α)∥ ≤ n * ∥(1
 | (n + 1) := by { rw [n.cast_succ, n.cast_succ, add_mul, one_mul],
                   exact norm_add_le_of_le (nat.norm_cast_le n) le_rfl }
 
+end semi_normed_group
+
+section normed_group
+
+/-- Construct a normed group from a translation invariant distance -/
+def normed_group.of_add_dist [has_norm α] [add_comm_group α] [metric_space α]
+  (H1 : ∀ x:α, ∥x∥ = dist x 0)
+  (H2 : ∀ x y z : α, dist x y ≤ dist (x + z) (y + z)) : normed_group α :=
+{ dist_eq := λ x y, begin
+    rw H1, apply le_antisymm,
+    { rw [sub_eq_add_neg, ← add_right_neg y], apply H2 },
+    { have := H2 (x-y) 0 y, rwa [sub_add_cancel, zero_add] at this }
+  end }
+
+/-- A normed group can be built from a norm that satisfies algebraic properties. This is
+formalised in this structure. -/
+structure normed_group.core (α : Type*) [add_comm_group α] [has_norm α] : Prop :=
+(norm_eq_zero_iff : ∀ x : α, ∥x∥ = 0 ↔ x = 0)
+(triangle : ∀ x y : α, ∥x + y∥ ≤ ∥x∥ + ∥y∥)
+(norm_neg : ∀ x : α, ∥-x∥ = ∥x∥)
```

Comment:
Can this extend `semi_normed_group.core`?

Reply:
Done.

Reply:
Not so fast:
```
structure normed_group.core (α : Type*) [add_comm_group α] [has_norm α]
  extends semi_normed_group.core α : Prop :=
(norm_eq_zero_if : ∀ x : α, ∥x∥ = 0 → x = 0)
```
works, but now the fields are different, so I have to modify it every time it is used (not that much, so it's doable). What I currently did is to keep this definition and to provide `normed_group.core.to_semi_normed_group.core` by hand, so it can be used in `normed_group.of_core`.